File: dca-05-02-2019/producer/src/test/java/com/example/producer/UserPersistenceTests.java
package com.example.producer;

import org.assertj.core.api.Assertions;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.reactivestreams.Publisher;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.data.mongo.DataMongoTest;
import org.springframework.test.context.junit.jupiter.SpringExtension;
import reactor.core.publisher.Flux;
import reactor.test.StepVerifier;

import java.util.UUID;


@DataMongoTest
@ExtendWith(SpringExtension.class)
public class UserPersistenceTests {

    @Autowired
    private UserRepository repo;

    @Test
    public void testShouldSaveFind() {
        DemoUser user = new DemoUser(UUID.randomUUID(), "PureLife");

        Flux<DemoUser> userStream = repo
                .save(user)
                .thenMany(repo.findAll());

        StepVerifier
                .create(userStream)
                .assertNext(it ->
                        Assertions
                                .assertThat(it)
                                .as("a demo user has state")
                                .isNotNull()
                                .hasNoNullFieldsOrProperties()
                )
                .verifyComplete();
    }

    @Test
    public void testShouldStream() {
        DemoUser user = new DemoUser(UUID.randomUUID(), "PureLife");

        Publisher userFlux = Flux
                .just(user)
                .repeat(1);

        StepVerifier
                .create(userFlux)
                .expectSubscription()
                .assertNext(it ->
                        Assertions
                                .assertThat(it)
                                .as("a demo user has state")
                                .isNotNull()
                                .hasNoNullFieldsOrProperties()
                )
                .assertNext(it ->
                        Assertions
                                .assertThat(it)
                                .as("a demo user has state")
                                .isNotNull()
                                .hasNoNullFieldsOrProperties()
                )
                .expectComplete()
                .verify();

    }

}

File: capetown-04-12-19/capetown-producer/src/test/kotlin/com/demo/capetown/producer/UserPersistenceTests.kt
package com.demo.capetown.producer

import org.assertj.core.api.Assertions
import org.hamcrest.MatcherAssert
import org.hamcrest.Matchers
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.autoconfigure.data.mongo.DataMongoTest
import org.springframework.test.context.junit.jupiter.SpringExtension
import reactor.core.publisher.Mono
import reactor.test.StepVerifier
import java.time.Instant
import java.util.*

@DataMongoTest
@ExtendWith(SpringExtension::class)
class UserPersistenceTests {

    @Autowired
    lateinit var repo: UserRepo

    @Test
    fun `should stream user and verify`() {
        val userPublisher = Mono
                .just(DemoUser(UUID.randomUUID(), "Mario", Instant.now()))

        StepVerifier
                .create(userPublisher)
                .expectSubscription()
                .assertNext {
                    Assertions
                            .assertThat(it)
                            .isNotNull
                            .hasNoNullFieldsOrProperties()
                            .hasFieldOrPropertyWithValue("name", "Mario")
                }
                .verifyComplete()
    }

    @Test
    fun `should persist and find`() {
        val persistencePublisher = repo
                .save(DemoUser(UUID.randomUUID(), "Mario", Instant.now()))
                .thenMany(repo.findAll())

        StepVerifier
                .create(persistencePublisher)
                .expectSubscription()
                .assertNext {
                    MatcherAssert.assertThat("DemoUser has State", it,
                            Matchers.allOf(
                                    Matchers.notNullValue(),
                                    Matchers.hasProperty("name", Matchers.equalTo("Mario"))
                            ))
                }
                .expectNextCount(0)
                .verifyComplete()

    }
}